### ISSUE TITLE
Need a formal way to register element resources

### ISSUE THREAD

**Issue by sfmskywalker** [OrchardCMS/Orchard#6221]
Currently there is no formal way for custom elements to register their resources (scripts and styles). Instead, the module would need to implement `IShapeTableProvider` and attach a callback for the `OnDisplaying` event of the `EditorTemplate` shape. This works when the layout editor is being rendered by the `LayoutPartDriver`, but fails when custom modules reuse the layout editor by rendering the `LayoutEditor` view model directly using the `Html.Editor` HTML helper.

One suggestion is to turn the `LayoutEditor` view model into a shape and implement a default shape table provider that itself triggers an event on a formalized event handler, which custom modules can implement to register element resources.

Another approach could be to automatically include resources based on a convention, similar to the way Workflows is doing that. the downside is that we would get a lot of 404s, since not all elements will need to provide resources.

So the ultimate solution I think is to enable element classes / descriptors to specify the resources they want included. These resources need to be rendered at the same time the layout editor is rendered, since the elements themselves are rendered using AJAX.

**Comment by Piedone** [comment]
Also wanted to add that this is the same issue as with Workflows and the Workflows solution that blindly includes files is far from ideal.

**Comment by sfmskywalker** [comment]
Yeah, true. Could be solved with the same suggested approach (the one I labeled to be the ultimate one ;))
